From 98f489a3b74f55c887f0751fb6644e109a4df1ba Mon Sep 17 00:00:00 2001
From: SilverPlate3 <arielsilver77@gmail.com>
Date: Sun, 25 Feb 2024 09:54:52 +0200
Subject: [PATCH] add /sys/devices/platform/pcspkr/id

---
 drivers/input/misc/pcspkr.c | 35 +++++++++++++++++++++++++++++++++++
 1 file changed, 35 insertions(+)

diff --git a/drivers/input/misc/pcspkr.c b/drivers/input/misc/pcspkr.c
index 897854fd245f..48f9ba522dfd 100644
--- a/drivers/input/misc/pcspkr.c
+++ b/drivers/input/misc/pcspkr.c
@@ -14,12 +14,41 @@
 #include <linux/platform_device.h>
 #include <linux/timex.h>
 #include <linux/io.h>
+#include <linux/sysfs.h>
 
 MODULE_AUTHOR("Vojtech Pavlik <vojtech@ucw.cz>");
 MODULE_DESCRIPTION("PC Speaker beeper driver");
 MODULE_LICENSE("GPL");
 MODULE_ALIAS("platform:pcspkr");
 
+#define EUDYPTULA_ID "f3e7d7d3e3e7"
+#define EUDYPTULA_ID_LEN 13
+#define EUDYPTULA_ID_LEN_EXCLUDING_NULL EUDYPTULA_ID_LEN - 1
+
+static ssize_t id_show(struct device *dev, struct device_attribute *attr, char *buf)
+{
+    return sysfs_emit(buf, "%s\n", EUDYPTULA_ID);
+}
+
+static ssize_t id_store(struct device *dev, struct device_attribute *attr,
+                        const char *buf, size_t count)
+{
+    if(count != EUDYPTULA_ID_LEN_EXCLUDING_NULL)
+        return -EINVAL;
+
+    char input[EUDYPTULA_ID_LEN_EXCLUDING_NULL] = {0};
+    sscanf(buf, "%s", input); 
+
+    if (strcmp(input, EUDYPTULA_ID) != 0) 
+    {
+        return -EINVAL;
+    }
+
+    return count;
+}
+
+static DEVICE_ATTR(id, S_IRUGO | S_IWUSR, id_show, id_store);
+
 static int pcspkr_event(struct input_dev *dev, unsigned int type,
 			unsigned int code, int value)
 {
@@ -90,6 +119,12 @@ static int pcspkr_probe(struct platform_device *dev)
 		return err;
 	}
 
+	err = sysfs_create_file(&dev->dev.kobj, &dev_attr_id.attr);
+	if (err) {
+        input_free_device(pcspkr_dev);
+        return err;
+    }
+
 	platform_set_drvdata(dev, pcspkr_dev);
 
 	return 0;
-- 
2.25.1

